### DISCUSSION TITLE
[FR] Présentation

### DISCUSSION THREAD

**Discussion by GPh83** [GPh83/MailServerMonitor#2] [General]
# Introduction

Il vous est certainement arrivé de constater des indisponibilités ou des lenteurs de votre boite email.

Quand c’est occasionnel on patiente, quand c’est récurrent on contacte le support client de l’hébergeur. Le dialogue avec le support des différents opérateurs/hébergeurs est souvent compliqué sur ce genre de problème de qualité et d’évènements intermittents.

Afin de pouvoir dialoguer avec eux et avoir une vue objective, il faut quantifier et avoir des données fiables. D’où le développement de ce petit outil qui teste les connexions à un ou plusieurs serveurs SMTP/IMAP et génère un historique.

# Principe

On interroge le serveur et on journalise le succès et le temps de réponse. Concrètement, cela envoie un e-mail en SMTP à sa propre adresse et ensuite ça l’efface en IMAP. On valide donc le fonctionnement (liaison et authentification) et on mesure les temps pour les deux protocoles.

L’outil se contente de générer les données sous forme d’un historique au format CSV. Chacun pourra ainsi l’exploiter avec un tableur (Libre Office Calc, Excel, OnlyOffice, Calligra Sheets, etc.) pour obtenir ce qui l’intéresse (temps de réponse moyen/max/min) ou taux d’erreur…

# Usage

C’est un outil en ligne de commande qui possède un fichier de configuration en JSON.

Selon la configuration, une fois lancé il fonctionne de manière perpétuelle en interrogeant toutes les X minutes. Si on veut que cela fonctionne automatiquement, il faut le déclarer en service sous Linux ou dans une tache planifiée au démarrage de la machine sous Windows.

Les commandes sont les suivantes :

    -a : ajout d’une boite à lettre à tester
    -l : liste les boites configurées
    -d : suppression d’une boite configurée
    -u : permet de lancer une exécution unique même si la config indique un mode perpétuel.

# Configuration

Le fichier JSON est crée automatiquement lors de la première utilisation.
- `AskMin `: interrogation toutes les X minutes. Par défaut =0=Interrogation unique
- `CSVName `: le nom du fichier de sortie. Par défaut="[MailServerMonitor](https://github.com/GPh83/MailServerMonitor).csv".

Pour chaque boite :
- `eMail `: l’e-mail à tester en envoi/réception
- `TimeOutMs `: Time out in millisecondes, par défaut =30000
- `ServerName `: le nom du serveur ou son IP
- `Login `: nom d’utilisateur pour l’authentification
- `EncryptedPassword `: utiliser la commande -a pour le créer
- `ImapPort `: le port IMAP, par défaut =993
- `ImapSSL `: pour utiliser SSL/TLS, par défaut =true
- `SmtpPort `: le port SMTP, par défaut =465
- `SmtpSSL `: pour utiliser SSL/TLS, par défaut=true

# Exploitation

L’historique contient une ligne par test pour chaque serveur configuré.
Le fichier dont le nom est paramétrable contient en entête les noms des colonnes.

On y trouve :
- le nom/ip du serveur `ServerName`
- l'e-mail de la boite `eMail`
- la date et heure de l’évènement `EventDateTime`
- si l'envoi et la réception on abouti `SendRecieveOk`
- si le serveur SMTP fonctionne `StmpConnected`
- le temps de réponse du serveur SMTP `SmtpResponseTimeMs`
- le message d’erreur associé au SMTP en cas d’erreur `SmtpError`
- pareil pour l’IMAP `ImapConnected` `ImapResponseTimeMs` `ImapError`
- le nombre de messages effacés lors de cette session `DeletedCount`

# Conclusion

C’est du .Net 6 donc utilisable sous Windows et Linux.
Si ça peut vous servir, le monde de l’Open Source et ma modeste contribution ne seront pas inutiles.

**Comment by francois-dorin** [comment]
Hello (Francesco de linuxfr ;) )

Est-ce que le choix de la licence (GPL) est un choix réfléchis ou simplement "parce qu'il fallait en mettre une" ?

J'ai déjà un outil de monitoring dans l'interface d'administration de mon client et si je dois intégrer cet outil, cela sera en tant que lib et non en tant qu'exécutable. Mais du coup, la GPL me bloque car incompatible.

**Reply by GPh83** [reply]
C'était un choix (pas spécifique à ce projet mais antérieur).
J'avais hésité à l'époque avec l'Apache 2, si cela peut t'arranger pourquoi pas.

**Reply by francois-dorin** [reply]
Apache 2 serait effectivement beaucoup mieux, car il n'y a pas la viralité de la GPL.

Après, mais là, ce n'est qu'un avis perso, je suis plutôt pour la MIT. Enormément de bibliothèque en C# sont sous MIT. Cela évite d'avoir 36000 licences différentes au sein d'un projet quand on ajoute des paquets Nuget.

Maintenant, je me pose quand même la question entre partir du projet existant ou en refaire un autre, pour les raisons suivantes :
- le projet est petit donc c'est tout à fait faisable
- il est pensé synchrone et non asynchrone (donc pour tester l'envoi et la réception depuis des comptes différents cela va nécessiter pas mal de refactoring)
- il est pensé pour une seule boite mail (donc pour tester l'envoi et la réception depuis des comptes différents, cela va nécessiter pas mal de refactoring (oui je sais, l'argument sert 2x !))
- le fichier de configuration sera changé de manière qui sera très certainement non rétro-compatible
- l'algorithme de chiffrement est à changer (le 3DES est considéré comme obsolète), et sera à remplacer par un algorithme plus robuste comme un AES
- la clé de chiffrement ne devra pas être en dur dans le code (on manipule des données sensibles : accès à des boites mails !) mais devrait plutôt être machine et utilisateur dépendant (utilisation du magasin de certificats, variables d'environnement, etc...)
- du point précédent, il faudrait inclure une procédure d'export/import
- pas de tests existants

Maintenant, quoi qu'il arrive, j'ai besoin du go de mon client avant de me lancer dans cette aventure (que ce soit une évolution du projet existant ou un nouveau).

**Reply by GPh83** [reply]
La licence est changée.
Après à toi de voir, c'est sur que si tu dois presque tout refaire ...

Un certain nombre de choses sont déjà améliorée mais pas encore push.
A suivre donc.
